### PR TITLE
Reject Builds with unresolved image references

### PR DESCRIPTION
**Pull request by mmilata**
BZ: https://bugzilla.redhat.com/show_bug.cgi?id=1384973
GitHub issue: #11324

### CODE REVIEW

File: pkg/build/api/validation/validation.go [hunk 1]
```diff
@@ -29,6 +29,32 @@ func ValidateBuild(build *buildapi.Build) field.ErrorList {
...
+	// Check that all source image references are of DockerImage kind.
+	strategyPath := field.NewPath("spec", "strategy")
+	switch {
+	case build.Spec.Strategy.SourceStrategy != nil:
+		allErrs = append(allErrs, validateBuildImageReference(&build.Spec.Strategy.SourceStrategy.From, strategyPath.Child("sourceStrategy").Child("from"))...)
+		allErrs = append(allErrs, validateBuildImageReference(build.Spec.Strategy.SourceStrategy.RuntimeImage, strategyPath.Child("sourceStrategy").Child("runtimeImage"))...)
+	case build.Spec.Strategy.DockerStrategy != nil:
+		allErrs = append(allErrs, validateBuildImageReference(build.Spec.Strategy.DockerStrategy.From, strategyPath.Child("dockerStrategy").Child("from"))...)
+	case build.Spec.Strategy.CustomStrategy != nil:
+		allErrs = append(allErrs, validateBuildImageReference(&build.Spec.Strategy.CustomStrategy.From, strategyPath.Child("customStrategy").Child("from"))...)
```
Comment (liggitt) [inline pkg/build/api/validation/validation.go]: do we know all the `from` kinds a custom strategy can resolve? could someone implement a custom strategy that resolved other kinds?

Comment (mmilata) [inline pkg/build/api/validation/validation.go]: My understanding is that custom strategy is implemented by means of docker image, the one `from` points to, so I'd say no. IOW the field is useful for build controller, not for the strategy itself.

The `from` field of custom strategy is used in this function: https://github.com/openshift/origin/blob/v1.4.0-alpha.0/pkg/build/controller/strategy/custom.go#L24 You can notice that the `kind` is not checked.

**Comment by mmilata** [comment]
@csrwng would you mind taking a look?

**Comment by csrwng** [comment]
LGTM [test]

**Comment by csrwng** [comment]
https://github.com/openshift/origin/issues/10988
[test]

**Comment by csrwng** [comment]
@mmilata looks like you need to fix the build admission integration test. It uses the same strategy for both builds and build configs. The strategy points to an image stream:
https://github.com/openshift/origin/blob/master/test/integration/build_admission_test.go#L269

**Comment by mmilata** [comment]
[test]

**Comment by csrwng** [comment]
[merge]

**Comment by mmilata** [comment]
I've modified the test to use `DockerImage`s instead of `ImageStreamTag`s as the test functionality doesn't seem to depend on the latter.

**Comment by liggitt** [comment]
this tightens validation, right... what is the impact if someone has a kind set that is now disallowed?

**Comment by bparees** [comment]
> this tightens validation, right? what is the impact if someone has a kind stored in etcd today that is now disallowed?

that build would almost certainly already be in a failed state.

if they managed to have a build in the new state when they do this upgrade and change, the build will end up failed at some point thereafter.

**Comment by liggitt** [comment]
> the build will end up failed

can the build status be updated if there is a validation error in the build?

**Comment by bparees** [comment]
> can the build status be updated if there is a validation error in the build?

i suppose not. i'm pretty sure i don't care about this edge case.

**Comment by bparees** [comment]
(given the lack of bugs reported by people trying to create builds in this way, it's not something people are doing.)

**Comment by liggitt** [comment]
ok

**Pull request merged**